File: labs/lab4/task1.4_analyze.py
#!/usr/bin/env python3
"""Task 1.4: SBOM Analysis and Extraction. Run from labs/: python lab4/task1.4_analyze.py"""
import json
from pathlib import Path
from collections import Counter

BASE = Path(__file__).parent
SYFT = BASE / "syft" / "juice-shop-syft-native.json"
TRIVY = BASE / "trivy" / "juice-shop-trivy-detailed.json"
OUT = BASE / "analysis" / "sbom-analysis.txt"
LIC = BASE / "syft" / "juice-shop-licenses.txt"

with open(SYFT, encoding="utf-8") as f:
    syft = json.load(f)
with open(TRIVY, encoding="utf-8") as f:
    trivy = json.load(f)

lines = []

# Syft package counts
syft_types = Counter(a.get("type", "unknown") for a in syft.get("artifacts", []))
lines.extend(["=== SBOM Component Analysis ===", "", "Syft Package Counts:"])
for t, c in syft_types.most_common():
    lines.append(f"     {c} {t}")
lines.append("")

# Trivy package counts (os-pkgs + lang-pkgs only)
trivy_types = Counter()
for r in trivy.get("Results", []):
    cls = r.get("Class", "")
    if "os-pkgs" not in cls and "lang-pkgs" not in cls:
        continue
    key = f"{r.get('Target', 'Unknown')} - {r.get('Type', 'unknown')}"
    trivy_types[key] += len(r.get("Packages", []))
lines.extend(["Trivy Package Counts:"])
for k, c in trivy_types.most_common():
    lines.append(f"     {c} {k}")
lines.append("")

# License Analysis
lines.extend(["=== License Analysis ===", "", "Syft Licenses:"])
syft_lic = Counter()
for a in syft.get("artifacts", []):
    for l in a.get("licenses", []):
        syft_lic[l.get("value", l)] += 1
for lic, c in syft_lic.most_common():
    lines.append(f"     {c} {lic}")
lines.append("")

lines.append("Trivy Licenses (OS Packages):")
trivy_os = Counter()
for r in trivy.get("Results", []):
    if "os-pkgs" in (r.get("Class") or ""):
        for p in r.get("Packages", []):
            for lic in p.get("Licenses", []):
                trivy_os[lic] += 1
for lic, c in trivy_os.most_common():
    lines.append(f"     {c} {lic}")
lines.append("")

lines.append("Trivy Licenses (Node.js):")
trivy_node = Counter()
for r in trivy.get("Results", []):
    if "lang-pkgs" in (r.get("Class") or ""):
        for p in r.get("Packages", []):
            for lic in p.get("Licenses", []):
                trivy_node[lic] += 1
for lic, c in trivy_node.most_common():
    lines.append(f"     {c} {lic}")

OUT.write_text("\n".join(lines) + "\n", encoding="utf-8")
print("Written:", OUT)

# Syft licenses file
lic_lines = ["Extracting licenses from Syft SBOM..."]
for a in syft.get("artifacts", []):
    if a.get("licenses"):
        vals = ", ".join(l.get("value", l) for l in a["licenses"])
        lic_lines.append(f"{a.get('name')} | {a.get('version')} | {vals}")
LIC.write_text("\n".join(lic_lines) + "\n", encoding="utf-8")
print("Written:", LIC)

File: labs/lab4/task1.4-analyze.js
#!/usr/bin/env node
/**
 * Task 1.4: SBOM Analysis and Extraction
 * Run from repo root: node labs/lab4/task1.4-analyze.js
 */
const fs = require('fs');
const path = require('path');

const BASE = path.join(__dirname);
const SYFT_JSON = path.join(BASE, 'syft', 'juice-shop-syft-native.json');
const TRIVY_JSON = path.join(BASE, 'trivy', 'juice-shop-trivy-detailed.json');
const OUT_ANALYSIS = path.join(BASE, 'analysis', 'sbom-analysis.txt');
const OUT_LICENSES = path.join(BASE, 'syft', 'juice-shop-licenses.txt');

const lines = [];

// Syft package counts
const syft = JSON.parse(fs.readFileSync(SYFT_JSON, 'utf8'));
const syftTypes = {};
for (const a of syft.artifacts || []) {
  const t = a.type || 'unknown';
  syftTypes[t] = (syftTypes[t] || 0) + 1;
}
lines.push('=== SBOM Component Analysis ===', '');
lines.push('Syft Package Counts:');
for (const [t, c] of Object.entries(syftTypes).sort((a, b) => b[1] - a[1])) {
  lines.push(`     ${c} ${t}`);
}
lines.push('');

// Trivy package counts
const trivy = JSON.parse(fs.readFileSync(TRIVY_JSON, 'utf8'));
const trivyTypes = {};
for (const r of trivy.Results || []) {
  const cls = r.Class || '';
  if (!cls.includes('os-pkgs') && !cls.includes('lang-pkgs')) continue; // skip secret, etc.
  const target = r.Target || 'Unknown';
  const type = r.Type || 'unknown';
  const key = `${target} - ${type}`;
  for (const _ of r.Packages || []) {
    trivyTypes[key] = (trivyTypes[key] || 0) + 1;
  }
}
lines.push('Trivy Package Counts:');
for (const [k, c] of Object.entries(trivyTypes).sort((a, b) => b[1] - a[1])) {
  lines.push(`     ${c} ${k}`);
}
lines.push('');

// License Analysis
lines.push('=== License Analysis ===', '');
lines.push('Syft Licenses:');
const syftLicenses = {};
for (const a of syft.artifacts || []) {
  for (const l of a.licenses || []) {
    const v = l.value || l;
    syftLicenses[v] = (syftLicenses[v] || 0) + 1;
  }
}
for (const [lic, c] of Object.entries(syftLicenses).sort((a, b) => b[1] - a[1])) {
  lines.push(`     ${c} ${lic}`);
}
lines.push('');

lines.push('Trivy Licenses (OS Packages):');
const trivyOsLicenses = {};
for (const r of trivy.Results || []) {
  if ((r.Class || '').includes('os-pkgs')) {
    for (const p of r.Packages || []) {
      for (const lic of p.Licenses || []) {
        trivyOsLicenses[lic] = (trivyOsLicenses[lic] || 0) + 1;
      }
    }
  }
}
for (const [lic, c] of Object.entries(trivyOsLicenses).sort((a, b) => b[1] - a[1])) {
  lines.push(`     ${c} ${lic}`);
}
lines.push('');

lines.push('Trivy Licenses (Node.js):');
const trivyNodeLicenses = {};
for (const r of trivy.Results || []) {
  if ((r.Class || '').includes('lang-pkgs')) {
    for (const p of r.Packages || []) {
      for (const lic of p.Licenses || []) {
        trivyNodeLicenses[lic] = (trivyNodeLicenses[lic] || 0) + 1;
      }
    }
  }
}
for (const [lic, c] of Object.entries(trivyNodeLicenses).sort((a, b) => b[1] - a[1])) {
  lines.push(`     ${c} ${lic}`);
}

fs.writeFileSync(OUT_ANALYSIS, lines.join('\n') + '\n');
console.log('Written:', OUT_ANALYSIS);

// Syft licenses file (package | version | licenses)
const licenseLines = ['Extracting licenses from Syft SBOM...'];
for (const a of syft.artifacts || []) {
  if (a.licenses && a.licenses.length > 0) {
    const vals = a.licenses.map(l => l.value || l).join(', ');
    licenseLines.push(`${a.name} | ${a.version} | ${vals}`);
  }
}
fs.writeFileSync(OUT_LICENSES, licenseLines.join('\n') + '\n');
console.log('Written:', OUT_LICENSES);
